Version for Armbian - Orange Pi

rozzimann96 opened the pull request:
Hi, do you plain make version for Orange Pi? For example Orange Pi PC+ with Armbian installed? 
I try to install but "cmake .." command every time end with error message.  Thx

dsafa22 committed:
删除默认代理

FD- committed:
Machine-translated Chinese comments and strings

FD- committed:
Moved dns registry to native code

FD- committed:
Removed java project

FD- committed:
Simplified audio processing for later supporting async audio decoders

FD- committed:
Fixed build on Linux

FD- committed:
Removed dependency on mDNSResponder, working build on Raspbian

FD- committed:
Minor improvement

FD- committed:
Started adding Raspberry Pi h264 renderer

FD- committed:
Fixed compilation

FD- committed:
Very first working video

FD- committed:
Handle large h264 data packets

FD- committed:
Use a video scheduler

FD- committed:
Properly setup tunnels

FD- committed:
Render in fullscreen

FD- committed:
Removed renderer setup, blank background needs different workaround

FD- committed:
Added black background for renderer

FD- committed:
Reduce wait for buffer

FD- committed:
Updated readme

FD- committed:
Updated readme

FD- committed:
Temporarily added benchmarking to h264 decryption

FD- committed:
Updated readme

FD- committed:
Added options for background and name

FD- committed:
Improved hw address

FD- committed:
Fixed compiler warnings

FD- committed:
Updated TODO

FD- committed:
Added fdk audio renderer

FD- committed:
Merged changes from upstream commit a2c3c8e

FD- committed:
Removed some debug output

FD- committed:
Updated readme

FD- committed:
Updated todo

FD- committed:
Fixed artifacts

FD- committed:
Improved interface

FD- committed:
Fixed compiler warnings

FD- committed:
Removed unused code

FD- committed:
Updated TODO

FD- committed:
Added GPLv3 license headers to files I authored

FD- committed:
Fixed readme format

FD- committed:
Fixed readme format

FD- committed:
Added option for disabling audio

FD- committed:
Added hint for disabling audio

FD- committed:
Removed redundant newlines from log output

FD- committed:
Updated TODO

FD- committed:
Fixed build

FD- committed:
Fix random connection losses

FD- committed:
Replaced tiny-aes and axtls with openssl

FD- committed:
Updated TODO

FD- committed:
Investigating playback stalls

FD- committed:
Investigating playback stalls

FD- committed:
Buffer wait is not causing playback stalls

FD- committed:
Hack for reducing latency

FD- committed:
Investigating closing hang

FD- committed:
Investigating closing hang

FD- committed:
Investigating video stalls

FD- committed:
Investigating video stalls

FD- committed:
Removed debug logging

FD- committed:
Updated TODO

FD- committed:
Disabled dumps

FD- committed:
Updated TODO

FD- committed:
Updated Readme

FD- committed:
Updated readme

FD- committed:
Use one global ntp thread

FD- committed:
Minor modifications

FD- committed:
Properly construct INFO response

FD- committed:
Added AirPlay protocol history section

FD- committed:
Added support for iOS 9

FD- committed:
Extended AirPlay history

FD- committed:
Updated license information

FD- committed:
Simplified bplist generation memory handling

FD- committed:
Updated README

FD- committed:
Updated README

FD- committed:
Update README.md

FD- committed:
Update README.md

FD- committed:
Create LICENSE

FD- committed:
Delete jni_raop_server.cpp

FD- committed:
Added history detail

FD- committed:
Fixed hw address formats in dnssd txt

FD- committed:
Added volume control

FD- committed:
Fixed teardown bug

FD- committed:
Update README.md

FD- committed:
Update README.md

FD- committed:
Update README.md

FD- committed:
Update README.md

FD- committed:
Added ntp clock sync

Minor function rename

FD- committed:
Cleaning up mirroring code

FD- committed:
Added rtp clock sync

FD- committed:
Adapted renderers to share clock

FD- committed:
Log video stalls

FD- committed:
Experimenting with manually updating the reference clock

FD- committed:
Fixed parsing of audio device

FD- committed:
Inject max_dec_frame_buffering into SPS NAL

FD- committed:
Added unsynchronised low-latency mode

FD- committed:
Updated version number

FD- committed:
Disabled audio dumping

FD- committed:
Added flag for controlling log level

FD- committed:
Use END_OF_FRAME instead of END_OF_NAL

FD- committed:
Adjusted log levels and messages

FD- committed:
Build libraries statically

FD- committed:
Reintroduced audio buffer

FD- committed:
Use receive timeouts where needed

FD- committed:
Fixed sporadic crypto bug

FD- committed:
Updated readme to reflect version 1.1

FD- committed:
Update README.md

FD- committed:
Log exact socket recv error

FD- committed:
Minor logging improvements

FD- committed:
Dont close connection on timeout

FD- committed:
Fixed issue #1

FD- committed:
Fixed issue #6

FD- committed:
Fixed indentation

FD- committed:
Fixed unknown stream type error message

FD- committed:
Use correct executable name in instructions

FD- committed:
Improved socket handling as per upstream commit 64b108f

FD- committed:
Show blank screen after connection ended

FD- committed:
Fixed issue #14

FD- committed:
Completed incomplete previous commit

FD- committed:
Added changes for version 1.2

FD- committed:
Cleaner fix for issue #14

FD- committed:
Fixed double-free in video renderer

FD- committed:
Update README.md

FD- committed:
Update README.md

FD- committed:
Fixed getting stuck destroying video renderer

FD- committed:
Added GCC version requirement

FD- committed:
Improved volume control

FD- committed:
Fixed memory leak in audio renderer

taoyuan committed:
Add `auto`, `hide` and `always` three background drawing mode support.

FD- committed:
Streamlined indentation

FD- committed:
Adapted background mode, added documentation and improved backwards-compatibility

FD- committed:
Updated documentation for new background modes

StevenHickson committed:
Updated to support rotation by a numbered flag

FD- committed:
Added rotation info to help

FD- committed:
Added documentation for rotation flag

pallas committed:
cmake: add install target

pallas committed:
renderers/fdk-aac: bump to 2.0.1

This commit syncs renderers/fdk-aac/ to

	https://github.com/mstorsjo/fdk-aac/releases/tag/v2.0.1

pallas committed:
lib/http_parser: bump from 2.8.1 to 2.9.3

This commit syncs lib/http_parser.{c,h} to

	https://github.com/nodejs/http-parser/releases/tag/v2.9.3

while preserving the RFC-2326 RTSP & RAOP code added by RPiPlay.

pallas committed:
lib/{dnssd,utils}: const-correctness nits

In `dnssd_get_airplay_txt`, `dnssd->TXTRecordGetBytesPtr` returns a
`const void*` and implicitly casts it to a `char*`.

In fact, no dnssd_get_* callers own memory and it should be considered
const.  As a side-effect, codify that `utils_parse_hex` does not modify the
buffer it is passed.

antimof committed:
first commit

antimof committed:
Update video_renderer_gstreamer.c

pallas committed:
dnssd: null terminate dnssd->name to prevent overrun in dnssd_register_raop

The use of strncat in dnssd_register_raop expects that dnssd->name is null
terminated but in dnssd_init only name_len bytes are being allocated.
Rather than trying to work around strncat semantics, just calloc an extra
byte to serve as the null terminator.

Signed-off-by: Derrick Lyndon Pallas <derrick@pallas.us>

pallas committed:
lib/plist: bump to 2.1.0

This commit syncs lib/plist/ to

	https://github.com/libimobiledevice/libplist/releases/tag/2.1.0

Signed-off-by: Derrick Lyndon Pallas <derrick@pallas.us>

pallas committed:
plist: time64/ctime64_r: localtime64_r can fail

In ctime64_r, the call to localtime64_r can fail.  If we don't check for
this and then call asctime64_r, the results are garbage.

Signed-off-by: Derrick Lyndon Pallas <derrick@pallas.us>

pallas committed:
raop/handlers: avoid strncmp overruns with length checks

In raop_handler_get_parameter and raop_handler_set_parameter, the buffers
passed to strncmp are not necessarily null terminated.  This is sometimes
causing a 1-byte overread.  Instead, check to make sure that the buffer is
large enough for the strncmp to succeed.  This also applies to one use of
strstr, but strnstr is a non-standard extension; therefore, convert to
strncmp + length check also.

Signed-off-by: Derrick Lyndon Pallas <derrick@pallas.us>

pallas committed:
main: fix logger_t leak at stop_server

This shows up as a leak in valgrind.  Even though the server is shutting
down (so the leak shouldn't matter in practice), clean up anyway.

Signed-off-by: Derrick Lyndon Pallas <derrick@pallas.us>

ysdoho committed:
Update dnssdint.h

Changed the AirPlay Bit field to disable Airplay Video protocol as well as HLS

pallas committed:
lib/dnssdint: disable rotation support

When rotation support is enabled, the client assumes the server knows how to
rotate video, which is not currently the case in this project for all
streams.  By disabling this bit, the client performs rotation before sending
video to the server.  Note that this does not keep devices from rotating
their screens, just where the rotation happens.

Signed-off-by: Derrick Lyndon Pallas <derrick@pallas.us>

pallas committed:
cmake: replace included libplist with system library

The system version is regularly updated, so use that instead.  Users now
need to

	sudo apt install libplist-dev

in order to build.

antimof committed:
Implement source timestamp synchronization

antimof committed:
Implement source timestamp synchronization

Enhance audio quality

antimof committed:
Update audio_renderer_gstreamer.c

antimof committed:
Update video_renderer_gstreamer.c

antimof committed:
Update video_renderer_gstreamer.c

antimof committed:
Update audio_renderer_gstreamer.c

antimof committed:
Check availability of necessary Gstreamer plugins

antimof committed:
Check availability of necessary Gstreamer plugins

pallas committed:
renderers: do not block waiting for buffers

If we can't get an input buffer from ilclient, there is something wrong.
Therefore, do not block waiting for one.  Instead, bail out to prevent
hanging forever.

Signed-off-by: Derrick Lyndon Pallas <derrick@pallas.us>

pallas committed:
renderers/audio: report audio delay

This is identical to the debug logging in the video renderer.

Signed-off-by: Derrick Lyndon Pallas <derrick@pallas.us>

pallas committed:
renderers: reset timing on excessive delay

In my experience, the delay number is usually a small negative value.  When
it gets too large in the positive direction, the system will hang waiting to
render the buffer.

It is not clear to me why the delay is becoming excessive but there are a
couple of reliable ways I've found to reproduce it.  With an iPhone 7+, I

	1. mirrored the phone,
	2. started some music,
	3. put the phone to sleep,
	4. after 15 seconds, woke the phone up.

Prior to this patch, the mirror would hang.  After this patch, it doesn't.

Note that resetting the clock on any delay is not helpful, as it leads to
jitter in streams that should legitimately be delayed.  In the bad case I
was seeing delays in excess of 1M and in the working case I was seeing
legitimate delays of around 50k, so it seemed reasonable to make the reset
threshold 100k.  This could probably be tuned better.

Signed-off-by: Derrick Lyndon Pallas <derrick@pallas.us>

pallas committed:
renderers/fdk-aac: clean up CMakeFile, only compile decoder

I wanted to avoid compiling the encoder since it takes a while and is not
used.  In the process, I decided to just list the source directories and use
a loop to define the target library.

Signed-off-by: Derrick Lyndon Pallas <derrick@pallas.us>

pallas committed:
Bump fdk-aac to pre-2.1.3 version

There hasn't been much volatility in fdk-aac recently, and any changes they
do make are being actively tracked.  Rather than wait for a real ex-AOSP
release, pull in what's there now and fix up later.

pallas committed:
renderers/video: do not leak h264_stream_t when injecting max_dec_frame_buffering

Signed-off-by: Derrick Lyndon Pallas <derrick@pallas.us>

pallas committed:
renderers/video: tweak delay reset logic

Since resetting the packet time affects whether the pipeline flushes on
teardown, only do so AFTER we have allocated a buffer and are sure we're
going to push data.  Also, require that the flush actually happens, i.e.
block on buffer allocation.

tripplet committed:
Suggest  parallel make

pallas committed:
lib/raop: Enable keepalive for TCP sockets

I ran into an issue where my Mac crashed while I was mirroring and the TCP
session to RAOP RTP mirror did not time out in a reasonable amount of time.
Since RPiPlay is interactive, enable keepalive after one minute and time out
after an additional minute.

Signed-off-by: Derrick Lyndon Pallas <derrick@pallas.us>

dougg3 committed:
Fix NTP sync functionality

The existing NTP synchronization has a few issues. It only waits 3
milliseconds for a response from the iOS device, which is too short of a
delay. Also, if it fails to send the NTP request for some reason, or it
fails to receive the NTP response in time, it stops attempting NTP
synchronization altogether.

This commit attempts to address these problems. The NTP response time
has been increased to 300 ms, and if we fail to send or receive an NTP
packet, we will try again next time. The NTP socket is flushed prior to
each request to ensure we don't read an older delayed response.

dougg3 committed:
Allow choosing renderer at build time

After merging the gstreamer renderers from UxPlay, this project needed a
method for choosing which renderer to build into the program. I opted
for supplying a CMake variable describing which renderer you wish to
use. For backwards-compatibility, it defaults to selecting the openmax
renderer that RPiPlay has historically used.

Because the gstreamer renderer doesn't support several of the options
that are supported by the openmax renderer, those options are removed
from the help output in a gstreamer build.

dougg3 committed:
Minor whitespace cleanup

dougg3 committed:
Update README with gstreamer build instructions

ShiverZm commented:
help，编译不过
![Uploading image.png…]()

ShiverZm commented:
![D1EF3351-95C8-4ede-9AA2-72AE6C404275](https://user-images.githubusercontent.com/26861023/89399806-8dcda000-d745-11ea-8b11-419dd9a7ce38.png)

dougg3 committed:
Refactor to allow compiling multiple renderers

This change makes it possible to select video and audio renderers at
runtime. It's also possible to mix and match different renderers for
audio versus video.

All available renderers are automatically compiled into the renderers
library. It should always succeed with at least the dummy renderer.

dougg3 committed:
Add explanation text for renderers in print_info

This adds an explanation for how to choose different renderers.

dougg3 committed:
Add videoparsersbad as a required GStreamer plugin

This plugin contains the H.264 parser required for video to work
properly. This is the easiest way to require that
gstreamer1.0-plugins-bad has to be installed.

dougg3 committed:
Formatting cleanup

Tried to match the rest of the project's code style.

dougg3 committed:
Update README.md to explain desktop build process

Also documents the -vr and -ar options.

dougg3 committed:
Fix warning on x86_64 Linux

%llu isn't correct on 64-bit Linux, where it should actually be %lu.
On 32-bit systems, %llu is correct. A simple way to work around this is
to use the PRIu64 format specifier introduced in C99.

dougg3 committed:
Fix formatting of help message

dougg3 committed:
Add support for horizontal/vertical flipping

This commit also adds rotation and flipping in the gstreamer renderer.

dougg3 committed:
Fix cross compiling

When cross-compiling, CMake fails to find the /opt/vc libraries because
they are really part of the cross compiler sysroot rather than your
build machine. This change instructs CMake to look in the cross-compile
sysroot for /opt/vc. It doesn't affect a build directly on the Pi.

dougg3 committed:
Refactor renderer init functions

The list of arguments to the renderer init functions was ballooning, so
this commit changes the organization so the video and audio renderer
init functions simply take a struct containing all of the configuration
info.

I also fixed a couple of small code style issues.

dougg3 committed:
Update README with arguments for flipping

bedrin committed:
Crash on "Got NULL buffer!" error

pallas committed:
lib: s/http_parser/llhttp/

The NodeJS project replaced http_parser with llhttp, which is safer, faster,
and easier to maintain than its predecessor.  The license is MIT for both
http_parser and llhttp.  This commit makes the switch for this project, but
depends on a small patch [1] to add the RTSP+RAOP verbs and allow the RTSP
protocol type, & will hopefully be pulled upstream.

Generated code based on llhttp v2.1.3 + patch.

[1] https://github.com/nodejs/llhttp/pull/66

pallas committed:
Merge pull request #167 from pallas/s-http-parser-llhttp

lib: s/http_parser/llhttp/

pallas committed:
Fixup Raspberry Pi h264 stream hack

The h264bitstream library is quite dangerous because it does not check
bounds on any of the parsed arrays.  However, it is quite useful to be able
to inject an optimized max_dec_frame_buffering into h264 streams on OMX
hardware to lower the rendering pipeline latency.

This change mashses up a couple of things to make h264bitstream usage safe.
First, we do not allocate buffers or the h264 structure until we are pretty
sure we can inject.  Second, we do not read the existing frame and modify
it; instead, we generate a brand new frame with the appropraite flags set
and splice it into the bistream.  Third, we do not cut off the preceding
frame, which could have theoretically happened although was hard to produce
in a practical setup.  Fourth, we only allocate a small modification buffer
and if the NAL would break it, which should not happen with the smaller NAL,
we bail out of the injection.

I've been running the change locally and the change does seem to help with
pipeline latency without having to jump through all the parse/unparse hoops.

pallas committed:
lib/llhttp: add MIT license

pallas committed:
lib/llhttp: bump to 2.2.0

This includes upstream support for RTSP/RAOP. [1]

[1] https://github.com/nodejs/llhttp/pull/68

pallas committed:
Merge pull request #177 from pallas/bump-llhttp-2-2-0

lib/llhttp: bump to 2.2.0

pallas committed:
raop/get_parameter: fix memory leak

pallas committed:
Merge pull request #172 from pallas/mysterious-future

Mysterious future

Commit:
Add plist 2.0 to lib/CMakeLists.txt

add the secondary target for LIBPLIST as plist-2.0, as needed for and tested on Fedora 33

Commit:
Added Fedora install notes, added make install

Also slightly reworked the layout of the Desktop Operating System section, to show instructions for multiple OS's better.

pallas committed:
README: s/http-parser/llhttp/

pallas committed:
cmake: fix install target

This change allows install from build directories that are not the source
directory.  PROGRAMS are part of the source directory while TARGETS/RUNTIME
are executibles generated by the build.  I've gotten this right in other
projects, just messed it up here.

tl;dr: s/PROGRAMS/TARGETS/

pallas committed:
renderers/fdk-aac: bump to pre-3.2.0

Same as before, we'll take what we can get.

jasLogic committed:
Use OpenSSL for elliptic curve crypto

This commit removes the curve25519-donna and ed25519 libraries and
replaces them with their OpenSSL implementations.

Two new macros were defined in `lib/pairing.h`:
- `X25519_KEX_SIZE` for the size of the keys in bytes
- `PAIRING_SIG_SIZE` for the length of the signature in bytes

These macros were implemented in `lib/raop_handlers.h` too.

Also, the `handle_error` function from `lib/crypto.c` was renamed to
`crypto_handle_error`, improved slightly and placed into the
header `lib/crypto.h`.
This way `lib/pairing.c` can call this function when an OpenSSL
error occurs.

pallas committed:
Merge pull request #186 from pallas/fix-parameter-leak

raop/get_parameter: fix memory leak

jasLogic committed:
Move elliptic curve crypto into crypto files

This commit moves all code responsible for the elliptic curve
cryptography into the `lib/crypto.c` file into seperate functions.
All of OpenSSLs data structures are wrapper in appropiate structs.

Also, the `crypto_handle_error` functions is renamed back to
`handle_error` and removed from the `lib/crypto.h` header.

pallas committed:
Merge pull request #187 from jasLogic/master

Use OpenSSL for elliptic curve crypto

pallas committed:
Set OpenSSL compatibility level to 1.1.0

In 6f51dd0 Use OpenSSL for elliptic curve crypto, all crypto became the
responsibility of OpenSSL.  Unfortunately, CMake was set to require 1.1.1 or
higher, but Stretch was on 1.1.0 meaning that RPiPlay suddenly starts to
break for older version of RaspberryPi OS.  I don't see anything that
actually requires 1.1.1, so search for 1.1.0 or better and set the
appropriate API compatibility level for OpenSSL.

This change has been tried on Buster, but I don't have a Stretch install so
I'm crossing my fingers here with the knowledge that OPENSSL_API_COMPAT has
been robust enough in my past experience.

Issue: #192

pallas committed:
Set default build type to Release

Most users want a Release build, so change the default build type to that.
Developers who want a Debug build can explicitly ask for it.

pallas committed:
OpenSSL s/1.1.0/1.1.1/

The commit that moved ECC into OpenSSL chose version 1.1.1, and I should
have stuck with that, since that's where ECC got EVP support.  Whoops!

pallas committed:
renderers/fdk-aac: bump to upstream a52114d

Same as before.  I don't actually understand fdk-aac versioning, so
reference the merge hash instead.

pallas committed:
lib/llhttp: bump to 4.0.0

pallas committed:
renderers/fdk-aac: bump to upstream 6a2eb8b

pallas committed:
lib/llhttp: bump to 6.0.2

pallas committed:
renderers/fdk-aac: bump to v2.0.2

SoapGentoo committed:
Do not force inlining of indirect functions

* A function called indirectly cannot be decorated
  with `__attribute((always_inline))`, as this is
  guaranteed to only work with direct calls:
    https://gcc.gnu.org/bugzilla/show_bug.cgi?id=63220#c1

Bug: https://bugs.gentoo.org/798045

pallas committed:
lib/llhttp: bump to 6.0.4

Actually a no-op since we're not wasm, but meh.

pallas committed:
lib/llhttp: bump to 6.0.6

pallas committed:
lib/crypto: fix up OpenSSL AES calls

Previously, we weren't calling Final after Update when decrypting (or
encrypting) via AES.  My assumption is that this is because it was failing
when due to a padding issue that caused Final to bail out.  This change
turns off padding for both encrypt/decrypt and adds the call to final.

We should probably consider having a static context that is reset after the
call rather than allocating/deallocating a new CTX every time.  I'm going to
punt on that for now though.

Issue: FD-/RPiPlay#283

fduncanh committed:
rpiplay.cpp: replace find_mac() with robust version valid for non-eth0 systems
tested on debian, ubuntu, OpenSUSE, macOS, FreeBSD

pallas committed:
renderers/fdk-aac: bump upstream 072f2cd

In order to make this easier in the future, use FDK-AAC's own CMakeLists.
An unfortunate side effect is that we're now compiling the encoder again, so
the build time is longer and the binary is larger.

pallas committed:
renderers/fdk-aac: do not overflow bwBorders
